src/topology/metric_space/pi_lp.lean

```diff
@@ -0,0 +1,369 @@
+
+/-- Two real exponents `p, q` are conjugate if they are `> 1` and satisfy the equality
+`1/p + 1/q = 1`. This condition shows up in many theorems in analysis, notably related to `L^p`
+norms. -/
+structure is_conjugate_exponent (p q : ℝ) : Prop :=
+(one_lt : 1 < p)
+(inv_add_inv_conj : 1/p + 1/q = 1)
+
+/-- The conjugate exponent of `p` is `q = p/(p-1)`, so that `1/p + 1/q = 1`. -/
+def conjugate_exponent (p : ℝ) : ℝ := p/(p-1)
+
+/- Register several non-vanishing results following from the fact that `p` has a conjugate exponent
+`q`: many computations using these exponents require clearing out denominators, which can be done
+with `field_simp` given a proof that these denominators are non-zero, so we record the most usual
+ones. -/
+lemma is_conjugate_exponent.pos {p q : ℝ} (h : p.is_conjugate_exponent q) :
+  0 < p :=
+lt_trans zero_lt_one h.one_lt
+
+lemma is_conjugate_exponent.ne_zero {p q : ℝ} (h : p.is_conjugate_exponent q) :
+  p ≠ 0 :=
+ne_of_gt h.pos
+
+lemma is_conjugate_exponent.sub_one_ne_zero {p q : ℝ}
+  (h : p.is_conjugate_exponent q) : p - 1 ≠ 0 :=
+sub_ne_zero_of_ne (ne_of_gt h.one_lt)
+
+lemma is_conjugate_exponent.one_div_pos {p q : ℝ} (h : p.is_conjugate_exponent q) :
+  0 < 1/p :=
+one_div_pos_of_pos h.pos
+
+lemma is_conjugate_exponent.one_div_ne_zero {p q : ℝ} (h : p.is_conjugate_exponent q) :
+  1/p ≠ 0 :=
+ne_of_gt (h.one_div_pos)
+
+lemma is_conjugate_exponent_iff {p q : ℝ} (h : 1 < p) :
+  p.is_conjugate_exponent q ↔ q = p/(p-1) :=
+begin
+  have A : 1/p + 1/q = 1 ↔ q = 1/(1-1/p) := calc
+    1/p + 1/q = 1 ↔ 1/q = 1-1/p : eq_sub_iff_add_eq'.symm
+    ... ↔ q = 1/(1-1/p) : by { simp only [one_div_eq_inv], rw inv_eq_iff, exact eq_comm },
+  split,
+  { assume H,
+    rw A.1 H.inv_add_inv_conj,
+    field_simp [H.ne_zero] },
+  { assume H,
+    refine ⟨h, _⟩,
+    rw H,
+    field_simp [ne_of_gt (lt_trans zero_lt_one h)] }
+end
+
+lemma is_conjugate_exponent_conjugate_exponent {p : ℝ} (h : 1 < p) :
+  p.is_conjugate_exponent (conjugate_exponent p) :=
+(is_conjugate_exponent_iff h).2 rfl
+
+namespace is_conjugate_exponent
+
+variables {p q : ℝ} (h : p.is_conjugate_exponent q)
+include h
+
+lemma conj_eq : q = p/(p-1) :=
+(is_conjugate_exponent_iff h.one_lt).1 h
+
+protected lemma symm : q.is_conjugate_exponent p :=
+{ one_lt :=
+    by { rw [h.conj_eq], exact one_lt_div_of_lt _ (sub_pos_of_lt h.one_lt) (sub_one_lt p) },
+  inv_add_inv_conj := by simpa [add_comm] using h.inv_add_inv_conj }
+
+end is_conjugate_exponent
+
+end real
+
+namespace finset
+/-- Hölder inequality: the scalar product of two functions is bounded by the product of their
+`L^p` and `L^q` norms when `p` and `q` are conjugate exponents. Version for sums over finite sets,
+with nonnegative functions. -/
+theorem sum_rpow_holder_of_nonneg {s : finset ι} {f g : ι → ℝ} {p q : ℝ}
+  (hpq : p.is_conjugate_exponent q) (hf : ∀ x ∈ s, 0 ≤ f x) (hg : ∀ x ∈ s, 0 ≤ g x) :
+  (∑ i in s, f i * g i) ≤ (∑ i in s, (f i)^p) ^ (1/p) * (∑ i in s, (g i)^q) ^ (1/q) :=
```

Comment:
I think that I know a shorter proof. I'll look into this later today. The idea is to use `real.young_inequality` instead of convexity.

Reply:
I pushed some proofs to https://github.com/leanprover-community/mathlib/blob/urkud-minkowskii/src/analysis/mean_inequalities.lean
I'll PR supporting changes to other files in a few minutes, and I can migrate the proofs to your branch tomorrow.

My main goal was to state explicitly that the `p`-norm is the maximum of scalar products with vectors from the `L^q`-norm unit ball, then just say that `max_x (f x + g x) ≤ max_x f x + max_x g x`. I also simplified the proof of Hölder inequality by normalizing both vectors; then pointwise application of Young inequality gives the required estimate.

Reply:
Sorry, I've accidentally deleted the branch. I'll restore it in a few hours.

Reply:
#3023 has supporting lemmas and `mean-ineqs-review` has proofs of Holder and Minkowskii for `real` and `nnreal`.